### ISSUE TITLE
IWebSocketConnectionFactory cannot be reimplemented

### ISSUE THREAD

**Issue by Shane32** [graphql-dotnet/server#672]
Because `IWebSocketConnectionFactory.CreateConnection` must return a `WebSocketConnection` (rather than an interface), and because `WebSocketConnection.Connect` is not marked as `virtual`, it is virtually impossible to write a custom implementation of `IWebSocketConnectionFactory`, making the interface useless.

**Comment by Shane32** [comment]
Here is sample code of what isn't possible because of the bad design:

```cs
    public class WebSocketConnectionFactory2<TSchema> : IWebSocketConnectionFactory<TSchema>
        where TSchema : ISchema
    {
        private readonly ILogger _logger;
        private readonly ILoggerFactory _loggerFactory;
        private readonly IGraphQLExecuter<TSchema> _executer;
        private readonly IEnumerable<IOperationMessageListener> _messageListeners;
        private readonly IDocumentWriter _documentWriter;

        public WebSocketConnectionFactory2(ILogger<WebSocketConnectionFactory<TSchema>> logger,
            ILoggerFactory loggerFactory,
            IGraphQLExecuter<TSchema> executer,
            IEnumerable<IOperationMessageListener> messageListeners,
            IDocumentWriter documentWriter)
        {
            _logger = logger;
            _loggerFactory = loggerFactory;
            _executer = executer;
            _messageListeners = messageListeners;
            _documentWriter = documentWriter;
        }

        //must return a class that inherits from WebSocketConnection (why?)
        public WebSocketConnection CreateConnection(WebSocket socket, string connectionId)
        {
            _logger.LogDebug("Creating server for connection {connectionId}", connectionId);

            var transport = new WebSocketTransport(socket, _documentWriter);
            var manager = new SubscriptionManager(_executer, _loggerFactory);
            var server = new SubscriptionServer(
                transport,
                manager,
                _messageListeners,
                _loggerFactory.CreateLogger<SubscriptionServer>()
            );

            return new WebSocketConnection2(transport, server, socket);
        }

        public class WebSocketConnection2 : WebSocketConnection
        {
            private readonly WebSocketTransport _transport;
            private readonly SubscriptionServer _server;
            private readonly WebSocket _socket;

            public WebSocketConnection2(
                WebSocketTransport transport,
                SubscriptionServer subscriptionServer,
                WebSocket socket) : base(transport, subscriptionServer)
            {
                _transport = transport;
                _server = subscriptionServer;
                _socket = socket;
            }

            //whoops!! cannot override Connect
            public override async System.Threading.Tasks.Task Connect()
            {
                try
                {
                    await _server.OnConnect();
                    await _server.OnDisconnect();
                    await _transport.CloseAsync();
                }
                finally
                {
                    //test bug fix here
                    _socket.Dispose();
                }
            }
        }
    }
```

**Comment by sungam3r** [comment]
@Shane32 This is just a trifle throughout the volume of all subscriptions problems. The whole project requires refactoring around connections/subscriptions/sockets/loggers/etc. Big work...

**Comment by Shane32** [comment]
Closed via #674
